### ISSUE TITLE
Las salidas solo deberían ser accesibles desde una dirección

### ISSUE THREAD

**Issue by zenieldanaku** [ludica-squamata/mano-gift#83]
Dado que la salida es un rect, según como esté puesta se puede acceder a otro mapa desde una posición ilógica. Por ejemplo, si se pisa una salida caminando de costado, el heroe aparece de costado (con la misma imagen) en el mapa siguiente, aún cuando la dirección sea ilógica.

**Comment by einacio** [comment]
No creo que en este caso lo relevante sea la Salida, sino más bien la
Entrada. Si tienen una dirección inicial va a ser más directo.

Limitar las Salidas se puede hacer desde mapa, no le veo sentido a
agregárselo lógicamente

**Comment by zenieldanaku** [comment]
A mi se me ocurrió hoy que, en el json del mapa (donde se define la
salida), simplemente sería cuestión de agregar una dirección. Luego, cuando
se carga la salida, si el mob que la pisa está en esa dirección, se lanza
el evento; si no, no.

**Comment by zenieldanaku** [comment]
Pienso que algo parecido deberíamos hacer con los interactuables también. Por ejemplo, la puerta, (¡que se puede abrir de espadas!) no debería ser posible de abrir sin estar _frente_ a ella.

**Comment by einacio** [comment]
si entiendo Interactivo, deberia ubicar el rect hacia adelante. tendria que
probarlo para entenderlo completamente, pero tengo la impresion de que no
me queda claro lo que hace y no se si esta ubicandolo correctamente.
(ademas de que tiene medidas fijas, asi que un sprite que no sea de 32x32
le puede quedar chico o tener mucho alcance)

por cierto, "interactuante" es actor en castellano (en el diccionario de la
rae es la segunda acepcion, participante en una accion). tambien se usa esa
palabra en ingles

**Comment by zenieldanaku** [comment]
Cuando dije "interactivo" en el issue me referia a los props con accion
(operable, agarrable, movible, etc). No al mob interactivo. Y recuerdo
haber puesto interactivo en lugar de interactuante porque no me acuerdo
dónde (creo que había sido el diccionario de la RAE, capaz una versión
vieja) me decía que no existía.

**Comment by zenieldanaku** [comment]
Sobre el último punto del primer post: Se me ocurre ahora que las salidas podrian indicar también algo más. Para el player, una salida es un cambio de mapas, pero para otros mobs, puede indicar un punto relevante para su AI. Por ejemplo, commoner podria saber que la salida en el interior de la casa cerca de la cual aparece es el centro de "su" casa, y huir ahi si fuera atacado por un monstruo, o cayera la noche. En este sentido, el heroe podria poder entrar a la casa siempre o, también, la salida podria estar "bloqueada". En ese caso solo actuaría como un indicador para la AI.

**Comment by zenieldanaku** [comment]
Sobre la idea del mapa de colisiones, se me ocurrió una nueva implementación: 
1) los datos iniciales permanecen en el archivo json.
2) el engine crea una imagen de tamaño igual al Chunk, un diccionario en blanco y una mascara en blanco del tamaño del chunk
3) por cada salida, pinta con un color aleatorio el area correspondiente a la salida
4) usando como key el color generado, loguea la información de la salida en el diccionario.
5) se setean los pixeles correspondientes a las salidas en 1

Luego, cuando los mobs caminen, se hace lo siguiente:
1) se detecta colisión usando la posición del mob contra la máscara de salidas
2) el mob pisa la salida.
3) se obtiene el color correspondiente al pixel de colisión de la imagen de salidas.
4) se usa el color como key para buscar la salida en el diccionario, sin recorrer la lista de salidas, e integrando el proceso a la detección de colisiones normal de los mobs

**Comment by zenieldanaku** [comment]
Lo anterior tiene tambien que ver con el issue #102 y el #115

**Comment by einacio** [comment]
no necesitas un segundo diccionario, la mascara es en rgb, asi que ya tenes
un diccionario ahi.

en vez de usar colores aleatorios, digamos que las salidas son los puntos
que tienen R=255, y en B pones el numero de salida, arrancando en 0. podes
indexar 255 salidas por mapa.
si el indice es GB, podes indexar 65536 salidas (probablemente te
alcanzaria para todas las del juego en un solo indice)

**Comment by einacio** [comment]
Perdon, 256/mapa

**Comment by zenieldanaku** [comment]
No. La mascara es un bits. 0 y 1 nomas. Y buscar colisión por PixelArray es mas lento que por mask

**Comment by zenieldanaku** [comment]
Tenes razón. No se me había ocurrido buscar por index (me quedé con la idea de que a las listas había que recorrelas)

**Comment by zenieldanaku** [comment]
Y te digo más: no es el mapa de colisiones sobre el que estamos pintando. Es un Surface distinto. Podemos usar los 4 canales, alpha incluido.

**Comment by zenieldanaku** [comment]
Sobre el último punto: se me ocurre que los otros mobs de hecho sí "cambien de mapa". Lo que pasaría es que se les agregaría a la definición del otro mapa (en memoria, no habría que modificar la especificación del mapa original), de modo que cuando el héroe pasare por esa salida, vería a los mobs que llegaron allí antes. Dichos mobs tendrían que tener alguna posición aleatoria (o predeterminada) en la que estar cuando el héroe llegase, de modo de no estar todos agolpados en la entrada.

**Comment by zenieldanaku** [comment]
Sin embargo, solo se actualizaría su posición. Al no estar presentes en la imagen, no hace falta renderearlos, ni actualizar su imagen u otros parámetros. Sólo habría que generar un diccionario con el nombre de mob como clave y su posición, aleatoria, como valor. De este modo, los otros mobs no respeterían la entrada, porque aparecerían en cualquier lugar del nuevo mapa.

**Comment by zenieldanaku** [comment]
Otra forma de determinar la posición del mob en el stage que el heroe no ve (o que verá cuando entre), es mediante AI. Si el mob se mete al bar para ponerse detrás de la barra, es porque es el bartender, y cuando el héroe entre al bar verá al mob en su puesto de trabajo. El punto donde el mob se para es relevante para ese mapa, lo cual significa que debería estar definido en su json.

**Comment by zenieldanaku** [comment]
> El punto donde el mob se para es relevante para ese mapa, lo cual significa que debería estar definido en su json
Pero eso va más allá de las salidas y el cambio de mapas, es un feature completamente nuevo que no está considerado siquiera.

**Comment by zenieldanaku** [comment]
Entonces, este issue quedaría algo asi:
- Todos los mobs desaparecen del mapa en el que figura la salida que pisaron.
- si la AI del mob es Controllable, entonces aparece en la entrada vincunada a la salida. si no, su posición es aleatoria.
- La posición de los mobs en el nuevo mapa no cambia hasta que la camara no se fija en ese nuevo mapa.
- Si el nuevo mapa tiene puntos marcados que son relevantes para la AI de algun mob, ese mob se moverá hacia ese punto.
